### Q1

Confidence level in the context of the chi-square statistic represents the probability of the fact that the observed data correspond to the fitted model for the given degrees of freedom.

First case

In [20]:
from scipy.stats import chi2

chi_square_value = 16

degrees_of_freedom = 13-3 #Because Quadratic Second Order

confidence_level = 1 - chi2.cdf(chi_square_value, degrees_of_freedom)

print("The Confidence level is {:.2%}".format(confidence_level))

The Confidence level is 9.96%


Second Case

In [21]:
chi_square_per_dof = 0.5

degrees_of_freedom_2 = 30

chi_square_value_2 = chi_square_per_dof * degrees_of_freedom_2

confidence_level_2 = 1 - chi2.cdf(chi_square_value_2, degrees_of_freedom_2)

print("The Confidence level is {:.2%}".format(confidence_level_2))

The Confidence level is 98.97%


Because confidence level is 98.97% which is really high. This means that the model fit the data too well. On the other hand, chi-square per degrees of freedom below 1 (0.5) might suggest that the error bars overestimated.

### Q2

HO=The two groups have the same mean.

H1=The two groups have different means.

In [22]:
import statistics

group1=[71, 59, 50, 51, 63, 60] 
group2=[46, 41, 68, 53, 47, 49, 35, 49, 54, 64]

mean_group1 = statistics.mean(group1)
mean_group2 = statistics.mean(group2)
print("The average of Group 1 is ",mean_group1)
print("The average of Group 2 is ",mean_group2)

The average of Group 1 is  59
The average of Group 2 is  50.6


In [23]:
std_group1 = statistics.stdev(group1)
std_group2 = statistics.stdev(group2)
print("The standard deviatian of Group 1 is {:.2f}".format(std_group1))
print("The standard deviatian Group 2 is {:.2f}".format(std_group2))

The standard deviatian of Group 1 is 7.82
The standard deviatian Group 2 is 9.86


In [24]:
n1=len(group1) 
n2=len(group2)
print(("Lenght of the group 1 is"),n1)
print(("Lenght of the group 2 is"),n2)

Lenght of the group 1 is 6
Lenght of the group 2 is 10


In [25]:
def S(n1,std_group1,n2,std_group2): 
  s_square=((n1-1)*(std_group1**2) + (n2-1)*(std_group2**2))/(n1+n2-2) 
  return s_square**(1/2) 

In [26]:
s=S(n1,std_group1,n2,std_group2) 
print("Standard Deviation: {:.2f}".format(s)) 

Standard Deviation: 9.18


In [27]:
def t_score(x_bar,y_bar,s,n1,n2): 
  res=(x_bar-y_bar)/(s*(((1/n1)+(1/n2))**(1/2))) 
  return res

In [28]:
t=t_score(mean_group1,mean_group2,s,n1,n2) 
print("t-score: {:.2f}".format(t))

t-score: 1.77


In [29]:
degrees_of_freedom=n1+n2-2 
print("Degree of freedom is {}".format(degrees_of_freedom))

Degree of freedom is 14


When I look at the table of significance level for two tails t table(In module resources in canvas)
Degrees of freedom is 14. 
Critical value at 0.05 significance level is 2.145

In [30]:
t_crit = 2.145

def t_hypothesis(t_actual,t_critic): 
  if t<=t_crit: 
    print("There is no enough evidence to reject the null hypothesis at 0.05 significance level") 
  else: 
    print("The null hypothesis can be rejected at 0.05 significance level") 
t_hypothesis(t,t_crit)

There is no enough evidence to reject the null hypothesis at 0.05 significance level


Therefore, they are not distinguishable.

### Q3

H0:Poisson distribution provides good fit

H1:Poisson distribution does not provide good fit

In [31]:
import math
number_of_events=[0,1,2,3,4,5,6] 
frequency=[320,387,189,76,22,4,2] 

alist=[] 
for (i,j) in zip(number_of_events,frequency): 
    alist.append(i*j) 

mean_num_events = sum(alist)/sum(frequency) 
print("Mean number of events: {}".format(mean_num_events)) 

Mean number of events: 1.113


In [32]:
def poisson(x,lam): 
    p=(((math.e)**(-lam))*(lam**x))/math.factorial(x) 
    return p 

pois_prob=[] 
pois_freq=[] 
print("Poisson probabilities & Expected frequencies for each event:") 

for i in number_of_events: 
    pp=poisson(i,mean_num_events) 
    pf=pp*sum(frequency) 
    print("{}= probability: {:.3f} & Expected frequency: {:.3f}".format(i,pp,pf)) 
    pois_prob.append(pp) 
    pois_freq.append(pf) 

Poisson probabilities & Expected frequencies for each event:
0= probability: 0.329 & Expected frequency: 328.572
1= probability: 0.366 & Expected frequency: 365.700
2= probability: 0.204 & Expected frequency: 203.512
3= probability: 0.076 & Expected frequency: 75.503
4= probability: 0.021 & Expected frequency: 21.009
5= probability: 0.005 & Expected frequency: 4.677
6= probability: 0.001 & Expected frequency: 0.867


In [33]:
n=7 
dof=n-2 
print("Degrees of freedom is {}".format(dof)) 

Degrees of freedom is 5


In [34]:
#Chi-square goodness of fit test
chi_square=0 
for (i,j) in zip(frequency,pois_freq): 
    chi_square += ((i-j)**2)/j 
print("chi-square: {:.3f}".format(chi_square)) 

chi-square: 4.125


When I look at the table of significance level for chi-square(In module resources in canvas)
Degrees of freedom is 5. 
Critical value at 0.05 significance level is 11.07

In [35]:
critical_val=11.07
def null_hypo_test(chi_square,critical_val): 
    if chi_square<=critical_val: 
        print("There is not enough evidence to reject null hypothesis at 0.05 significance level") 
    else: 
        print("The null hypothesis can be rejected at 0.05 significance level")

In [36]:
null_hypo_test(chi_square,critical_val) 

There is not enough evidence to reject null hypothesis at 0.05 significance level


Therefore, null hypothesis cannot be rejected.
Poisson distribution provides good fit to the data.

### Q4

In [37]:
# Given in the question
prob_covid = 36.4 / 100000  # P(COVID)
prob_pos_covid = 0.58  # P(Positive | COVID)
prob_false_pos = 0.011  # P(Positive | No COVID)

#Calculating the needed probabilities
prob_neg_covid = 1 - prob_pos_covid  # P(Negative | COVID)
prob_false_neg = 1 - prob_false_pos  # P(Negative | No COVID)
no_covid = 1 - prob_covid  # P(No COVID)

#Calculate P(Positive)
p_positive = (prob_pos_covid * prob_covid) + (prob_false_pos * no_covid)

#Calculate P(COVID | Positive) using Bayes
p_covid_given_positive = (prob_pos_covid * prob_covid) / p_positive

#Calculate P(Negative)
p_negative = (prob_neg_covid * prob_covid) + (prob_false_neg * no_covid)

#Calculate P(COVID | Negative) using Bayes
p_covid_given_negative = (prob_neg_covid * prob_covid) / p_negative

print("P(COVID | Positive) is {:.2%}".format(p_covid_given_positive))
print("P(COVID | Negative) is {:.3%}".format(p_covid_given_negative))


P(COVID | Positive) is 1.88%
P(COVID | Negative) is 0.015%


### Q6

In [38]:
import numpy as np
from scipy.stats import truncnorm

measured_flux = -0.4  # mean (measured flux) (wrong)
error_flux = 0.5  # standard deviation
minimum_flux = 0  # flux cannot be negative

#Calculating the lower bound for truncation
lower_bound = (minimum_flux - measured_flux) / error_flux  # Normalized lower bound

#Finding the 95% point in the truncated Gaussian.
upper_limit_flux = truncnorm.ppf(0.95, lower_bound, np.inf, loc=measured_flux, scale=error_flux)

round_upperlimit=round(upper_limit_flux,2)

print('Upper Limit is {}'.format(round_upperlimit))

Upper Limit is 0.75
